akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala

```diff
@@ -346,9 +349,12 @@ final class Split[T](decision: Split.SplitDecision, p: T ⇒ Boolean) extends Gr
             closeThis(handler, elem)
             handOver(handler)
           } else {
-            // Drain into the void
-            if (substreamCancelled) pull(in)
-            else substreamSource.push(elem)
+            if (substreamCancelled) {
+              if (eagerClose) completeStage()
```

Comment:
It is `onUpstreamFinish` which is signalled when the last element of a substream is consumed. There is no `onDownstreamFinish` on a port which has been closed, which is usually what happens after `onUpstreamFinish`. The only situations where a stage will get a `onDownstreamFinish` if the consumer does not want to read more. See http://doc.akka.io/docs/akka-stream-and-http-experimental/2.0.2/scala/stream-customize.html#Port_states__InHandler_and_OutHandler